File: suites/osprey-compute-kernels/reference/cases/wordfreq/wordfreq.hs
{-# LANGUAGE BangPatterns #-}

module Main (main) where

import Control.Monad (replicateM)
import Data.Array (accumArray, elems)
import Data.Char (isSpace)
import System.IO
  ( IOMode (ReadMode)
  , hGetChar
  , isEOF
  , withBinaryFile
  )
import Text.Read (readMaybe)

minstd :: Int
minstd = 16807

modulus :: Int
modulus = 2147483647

r1 :: Int -> Int
r1 x = (x * minstd) `mod` modulus

hashAt :: Int -> Int -> Int
hashAt s i = r1 (r1 (s + i))

vocab :: [String]
vocab = ["the", "quick", "brown", "fox", "jumps", "over", "lazy", "dog"]

readFirstLine :: IO String
readFirstLine = do
  eof <- isEOF
  if eof then pure "" else getLine

readSeed :: IO Int
readSeed = do
  line <- readFirstLine
  let m = maybe 0 id (readMaybe (filter (not . isSpace) line))
  if m == 0
    then pure 1
    else do
      cs <- withBinaryFile "/dev/urandom" ReadMode (replicateM 8 . hGetChar)
      let val = foldl (\acc c -> acc * 256 + fromEnum c) 0 cs
      pure (abs val `mod` 2147483646 + 1)

main :: IO ()
main = do
  seed <- readSeed
  let counts =
        elems (accumArray (+) 0 (0, 7)
          [(hashAt seed i `mod` 8, 1 :: Int) | i <- [0 .. 199999]])
      result = sum (zipWith (\k c -> (k + 1) * c) [0 ..] counts)
  print result

File: suites/osprey-compute-kernels/reference/cases/wordfreq/wordfreq.c
/* Word-frequency count over a tiny vocab — exercises hashing + counting. */
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#define MINSTD 16807
#define MODULUS 2147483647

static int64_t r1(int64_t x) { return (x * MINSTD) % MODULUS; }

static int64_t hash_at(int64_t s, int64_t i) { return r1(r1(s + i)); }

static int64_t read_seed(void) {
    char line[256];
    int64_t m = 0;
    if (fgets(line, sizeof line, stdin) != NULL) {
        char *end = NULL;
        long long parsed = strtoll(line, &end, 10);
        if (end != line) {
            m = (int64_t)parsed;
        }
    }
    if (m == 0) {
        return 1;
    }
    uint64_t val = 0;
    FILE *f = fopen("/dev/urandom", "rb");
    if (f != NULL) {
        if (fread(&val, sizeof val, 1, f) != 1) {
            val = 0;
        }
        fclose(f);
    }
    return (int64_t)(val % 2147483646) + 1;
}

int main(void) {
    const char *vocab[8] = {"the", "quick", "brown", "fox",
                            "jumps", "over", "lazy", "dog"};
    (void)vocab;
    int64_t seed = read_seed();

    int64_t counts[8] = {0};
    for (int64_t i = 0; i < 200000; i++) {
        counts[hash_at(seed, i) % 8] += 1;
    }

    int64_t result = 0;
    for (int64_t k = 0; k < 8; k++) {
        result += (k + 1) * counts[k];
    }

    printf("%lld\n", (long long)result);
    return 0;
}
